### ISSUE TITLE
Question: keys with multiple integers

### ISSUE THREAD

**Issue by PietropaoloFrisoni** [greg7mdp/parallel-hashmap#129]
Good morning,
I have a very simple question.

Let's say that I want to create a hash table and use keys identified by 3 integer values.
For example:

{1, 4, 5}  ---> 3
{3, 4, 6}  ---> 7
ecc...

Namely, each key can have only one single value associated to it, but the key itself is identified by 3 (ore more!) integers.

Is that possible to do something like this with these hash tables?

If the answer is yes, could you please provide me a super easy example which explains how to achieve this? 
Otherwise, if the answer is no, is there an easy way to do something like in c++, for example using the STL?

Thank you very much for your help and congratulations!

**Comment by greg7mdp** [comment]
Sure, it is not a problem, for example:

```
#include <iostream>
#include <array>
#include <parallel_hashmap/phmap.h>

struct MyKey
{
    friend size_t hash_value(const MyKey &k) 
    {
        return phmap::HashState().combine(0, k.key[0], k.key[1], k.key[2]);
    }

    bool operator==(const MyKey o) const { return key == o.key; }

    std::array<int, 3> key;
};

int main()
{
    using Hash = phmap::flat_hash_map<MyKey, int>;
    Hash h { 
        { { 1, 4, 5 }, 3 },
        { { 3, 4, 6 }, 7 }
    };

    h[MyKey{5, 5, 5}] = 11;

    std::cout << "Entry for key {3, 4, 6} is: " << h[MyKey{3, 4, 6}] << "\n";
 
    // and print it
    for (const auto& e : h)
        std::cout << e.second << "\n";
 
    return 0;
}
```

**Comment by PietropaoloFrisoni** [comment]
That's astounding!

Please just let me ask a couple of further questions:

1) My keys are identified by 9 integers and they "point to" a double value. If my key integers are very small (<12) can I reduce the table dimension (kB, MB etc.) by using "something else" instead of the standard "int" type for the key indices? If that's the case, can you please tell me what's the "something else" type?

2) When I load a map into the RAM:
phmap::BinaryInputArchive ar_in("...");
    h.phmap_load(ar_in);
how can I release the table and free the memory?

3) Is there an "easy way" to write on the h hash table above (the one that you wrote in your example) with several threads at the same time? 

4) I would like to cite this repository in my paper. Should I explicitly cite a paper with it?

Thank you very much!

**Comment by greg7mdp** [comment]
Thanks! Glad to be of help.

> If my key integers are very small (<12) 

If the integer's max value is less than 256, you can use `uint8_t`. so you could use `std::array<uint8_t, 9>`.
Actually even 4 bytes would be enough for each integer if 12 is the max value, so you probably could be a little more compact.

> how can I release the table and free the memory?

You can do h.clear() which will release the memory. Also memory is automatically released when the table gets out of scope. For example:

```
{
    phmap::flat_hash_map h;
    phmap::BinaryInputArchive ar_in("...");
    h.phmap_load(ar_in);

    // do something with h
}  //  <-- h gets out of scope, memory is released.
```

> Is there an "easy way" to write on the h hash table above (the one that you wrote in your example) with several threads at the same time?

you can use the `parallel_flat_hash_map` and provide a mutex as the last template parameter, which will allow safe insertion from multiple threads. For example:

```
using Map = phmap::parallel_flat_hash_map<std::size_t, std::size_t,
                                              std::hash<size_t>,
                                              std::equal_to<size_t>, 
                                              std::allocator<std::pair<const size_t, size_t>>, 
                                              4, 
                                              std::mutex>;
```

> I would like to cite this repository in my paper. Should I explicitly cite a paper with it?

Great! There is no paper to cite.

**Comment by PietropaoloFrisoni** [comment]
Amazing!

I have just a very last question before closing this. If I want to decide at runtime the path in which the hash tables will be stored, how can I safely use "BinaryOutputArchive" ?

For example, I tried to do something like this:

    (....) // definition of "filepath" (char *) and "integer_variable" (int)

    // build path for output table
    char filename[128];

    sprintf(filename, "..._%d...", integer_variable);  // of course the dots indicate generic characters that I don't copy here
    strcat(filepath, filename);

    phmap::BinaryOutputArchive ar_out(filepath);
    h.phmap_dump(ar_out);

But I obtain a very strange error at runtime. I am pretty sure that this is not the correct way to proceed, therefore could you please tell me how to achieve something like this?

**Comment by greg7mdp** [comment]
Hum, what is the type `h`? Also what is the error you are getting?
Just out of curiosity, what do these 9 small integers represent?

**Comment by PietropaoloFrisoni** [comment]
I will try and check. In the meantime, I can safely close this thread as you provided me all the informations I required.

Again, thank you very much for your time!

**Comment by greg7mdp** [comment]
There should no copies and it should be efficient. If you let me see your actual code I can confirm that it is as efficient as it should be.

**Comment by PietropaoloFrisoni** [comment]
Thank you man! 

I simply took your example and changed it a little bit:

struct MyKey
{
  friend size_t hash_value(const MyKey &k)
  {
    return phmap::HashState().combine(0, k.key[0], k.key[1], k.key[2], k.key[3], k.key[4], k.key[5], k.key[6], k.key[7], k.key[8]);
  }

  bool operator==(const MyKey o) const { return key == o.key; }

  std::array<uint8_t, 9> key;
};


and I use the table as:

  using Hash = phmap::flat_hash_map<MyKey, double>;
  Hash h{};

(...)

    hashed_tables_path_assigned = hashed_tables_path_assigned + "/Hashed_21j_symbols_dspin_" + std::to_string(dspin);
    phmap::BinaryInputArchive ar_in(&hashed_tables_path_assigned[0]);
    h.phmap_load(ar_in);

(...)


// key_21j is an array of integers.
    double  value = h[MyKey{key_21j[0], key_21j[1], key_21j[2], key_21j[3], key_21j[4],
                                  key_21j[5], key_21j[6], key_21j[7], key_21j[8]}];

(...)


The table is read from multiple threads, as it is my understanding that is thread-safe : )

**Comment by PietropaoloFrisoni** [comment]
Edit: I should have wrapped the code in angular brackets, sorry!

**Comment by greg7mdp** [comment]
I can't really tell from this, I'd need to see actual code.

> // key_21j is an array of integers.

Why not use MyKey instead?

> The table is read from multiple threads, as it is my understanding that is thread-safe : )

It is not safe to read it like that if insertions are still occurring.

**Comment by PietropaoloFrisoni** [comment]
I would like to show you the full code but it's simply too long and it would be a mess to copy and paste it here. I don't want to take so much time from you!

1) There are no insertions at this stage (read-only).

2) Since the key indices are inside a for loop, they change all the time. In a nutshell I have something like this:

```
for (int i1=0; i1 < N; i1++)
{
   for (int i2=0; i2 < N; i2++)
  {
     // other inner loops with i3, i4 etc.

     double value = h[MyKey{i1, i2, i3, i4, i5,
                                i6, i7, i8, i9}];

     (...)

  }
}
```

This obviously is not *exactly* my code but it gives an idea of the situation. Is there a more efficient way to proceed?

**Comment by greg7mdp** [comment]
This will be faster (note that there is no int ddeclaration after `for (`:
Also send me your MyKey struct declaration, and I'll send you a faster hash computation.

```
MyKey k;

auto& i0 = k.key[0];
auto& i1 = k.key[1];
auto& i2 = k.key[2];
auto& i3 = k.key[3];
auto& i4 = k.key[4];
auto& i5 = k.key[5];
auto& i6 = k.key[6];
auto& i7 = k.key[7];
auto& i8 = k.key[8];
auto& i9 = k.key[9];

for (i0=0; i0 < N; i0++)
{
    for (i1=0; i1 < N; i1++)
    {
        // other inner loops with i3, i4 etc.

        double value = h[k];

        (...)

    }
}
```

**Comment by greg7mdp** [comment]
Also, does the hash table contain an entry for all possible spin combinations? N^9? Because h[MyKey(...)] is creating an entry.

**Comment by PietropaoloFrisoni** [comment]
1) Yes, the hash table already contains all the possible spins combinations (which have been pre-computed before, this part is read-only)

2) This below is the "MyKey" struct:

```
struct MyKey
{
  friend size_t hash_value(const MyKey &k)
  {
    return phmap::HashState().combine(0, k.key[0], k.key[1], k.key[2], k.key[3], k.key[4], k.key[5], k.key[6], k.key[7], k.key[8]);
  }

  bool operator==(const MyKey o) const { return key == o.key; }

  std::array<uint8_t, 9> key;
};
```
The hash table is first loaded into memory:
```
using Hash = phmap::flat_hash_map<MyKey, double>;
Hash h{};

(...)

phmap::BinaryInputArchive ar_in(&hashed_tables_path_assigned[0]);
h.phmap_load(ar_in);
```

and then used as I have shown you before.

I don't know how to thank you!

**Comment by greg7mdp** [comment]
no problem :-).

Use this instead for MyKey (I just changed the `hash_value` function). It will be faster

```
struct MyKey
{
  friend size_t hash_value(const MyKey &k)
  {
      uint64_t x = *((uint64_t *)&key[0]);
      return phmap::HashState().combine(0, x, k.key[8]);
  }

  bool operator==(const MyKey o) const { return key == o.key; }

  std::array<uint8_t, 9> key;
};
```

**Comment by greg7mdp** [comment]
I just thought, if the hash table contains all the possible spin combinations, you could just use a std::vector instead.

**Comment by PietropaoloFrisoni** [comment]
Actually it's not that simple, as spins are constrained and are not exactly N^9 (the code I posted before it is just a trivial example in order to explain the situation).

That said, maybe a std::vector could be better. Can you please explain what you mean (I just started learning C++)?

Also, since (inside my inner loop) the hash table is "called" 4 times (each time with different spins order), I assume it would be appropriate to create just 4 MyKey objects at the beginning instead of creating an object every single time. 

Thank you again for the help!

**Comment by greg7mdp** [comment]
The loop as you sent it to me will insert all N^9 keys in the hash table. If N==12, that would be 5 billion pairs <MyKey, double>, probably using a lot of memory (100GB?).
If you don't want to always insert, you need to check if the key is here, so replace:

```
      double = h[k];
```

with

```
      auto it = h.find(k);
      if (it == h.end())
         continue;
      double value = *it;
      ...
```

**Comment by greg7mdp** [comment]
> maybe a std::vector could be better.

depends of how many values are in the table. You say it is less than N^9. How much less? half? 1/1000th?

**Comment by PietropaoloFrisoni** [comment]
1) The variables are constrained in the for loops(s) before reading the hash table, therefore it is not necessary to check before using it since the value certainly exists

2) So far the biggest hash table I generated is 400 Mb, but the biggest table I would like to generate should be 30 - 40 Gb (I will use a supercomputer). I actually don't know if there is a more efficient way to store everything by using a std::vector.

**Comment by greg7mdp** [comment]
> I actually don't know if there is a more efficient way to store everything by using a std::vector.

depends on how many values are in the table. You say it is less than N^9. How much less? half? 1/1000th?

**Comment by PietropaoloFrisoni** [comment]
It is not easy to answer, as I don't know it for sure. I just made rough estimates, but it depends on the value of N (it is a little bit tricky!). I think a realistic estimate is (N^9)/20. So far I get good results with these hash tables

**Comment by PietropaoloFrisoni** [comment]
I thought that in std:vector you are actually supposed to have just 1 index, but maybe there's a way to circumvent this

**Comment by greg7mdp** [comment]
what is the value of N? 12?

**Comment by greg7mdp** [comment]
Also, do you really need a double for your value, or could you use a float?

**Comment by PietropaoloFrisoni** [comment]
Since you seem really helpful and available, I write you (the most relevant parts of) the code below.

Notice that I didn't have time so far to implement the improvements that you suggested me

```
struct MyKey
{
  friend size_t hash_value(const MyKey &k)
  {
    return phmap::HashState().combine(0, k.key[0], k.key[1], k.key[2], k.key[3], k.key[4], k.key[5], k.key[6], k.key[7], k.key[8]);
  }

  bool operator==(const MyKey o) const { return key == o.key; }

  std::array<uint8_t, 9> key;
};
```


```
class Chain
{

(...)

private:
  // spins
  uint8_t ti_1, ti_2, ti_3, ti_4,
      ti_5, ti_6, ti_7, ti_8,
      ti_9, ti_10, ti_11, ti_12,
      ti_13, ti_14, ti_15, ti_16;

(...)

public:
  using Hash = phmap::flat_hash_map<MyKey, double>;
  Hash h{};
  int draw[16];

(...)

  // fictitious name, which gives an idea of ​​the situation
 double class_method_which_is_called_a_lot_of_times_() 
  {
    // prop_draw is an array with changes everytime this method is called,
   // so that spins need to be re-assigned every time

   // these below are useless copies that I still need to remove in order to optimize the code
    ti_1 = prop_draw[0];
    ti_2 = prop_draw[1];
    ti_3 = prop_draw[2];
    ti_4 = prop_draw[3];

    ti_5 = prop_draw[4];
    ti_6 = prop_draw[5];
    ti_7 = prop_draw[6];
    ti_8 = prop_draw[7];

    ti_9 = prop_draw[8];
    ti_10 = prop_draw[9];
    ti_11 = prop_draw[10];
    ti_12 = prop_draw[11];

    ti_13 = prop_draw[12];
    ti_14 = prop_draw[13];
    ti_15 = prop_draw[14];
    ti_16 = prop_draw[15];

   (...)

    for (tb1 = tb1_min; tb1 <= tb1_max; tb1 += 2)
    {

      tb2_min = abs(tb1 - dspin);
      tb2_max = tb1 + dspin;

      for (tb5 = tb5_min; tb5 <= tb5_max; tb5 += 2)
      {

        tb4_min = abs(tb5 - dspin);
        tb4_max = tb5 + dspin;

        for (tb2 = tb2_min; tb2 <= tb2_max; tb2 += 2)
        {
          for (tb4 = tb4_min; tb4 <= tb4_max; tb4 += 2)
          {

            tb3_min = max(abs(tb2 - dspin), abs(tb4 - dspin));
            tb3_max = min(tb2 + dspin, tb4 + dspin);

            for (tb3 = tb3_min; tb3 <= tb3_max; tb3 += 2)
            {

             (...)

              for (tpn1 = tpn1_min; tpn1 <= tpn1_max; tpn1 += 2)
              {

                tpn2_min = max(abs(tpn1 - dspin), abs(tb3 - dspin));
                tpn2_max = min(tpn1 + dspin, tb3 + dspin);

                for (tpn2 = tpn2_min; tpn2 <= tpn2_max; tpn2 += 2)
                {
 
                  (...)           

                 value_1= h[MyKey{ti_1, ti_2, ti_3, ti_4, tb1,
                                tb2, tb3, tpn1, tpn2}];

                  value_2= h[MyKey{ti_16, ti_15, ti_14, ti_13, tb5,
                                tb4, tb3, tpn1, tpn2}];

                  (...)

                } // tpn2
              }   // tpn1

                  (...)

              for (tps1 = tps1_min; tps1 <= tps1_max; tps1 += 2)
              {

                tps2_min = max(abs(tps1 - dspin), abs(tb3 - dspin));
                tps2_max = min(tps1 + dspin, tb3 + dspin);

                for (tps2 = tps2_min; tps2 <= tps2_max; tps2 += 2)
                {

                  value_3= h[MyKey{ti_8, ti_7, ti_6, ti_5, tb1,
                                tb2, tb3, tps1, tps2}];

                  value_4 = h[MyKey{ti_9, ti_10, ti_11, ti_12, tb5,
                                tb4, tb3, tps1, tps2}];
                   (...)

                } // tps2
              }   // tps1

            (...)

            } // tb3
          }   // tb4
        }     // tb2
      }       // tb5
    }         // tb1

  (...)

    return  (...) // something which depends on the 4 values
  }


```

If you don't have time to look at everything in detail, I understand perfectly.

I am 100% sure that there's a much more elegant way to write the algorithm, but C++ offers so many possibilities that I feel quite lost

**Comment by greg7mdp** [comment]
Hum, doesn't look like there is much to optimize. Just change hash_balue to: 

```
    friend size_t hash_value(const MyKey &k)
    {
        uint64_t x = *((uint64_t *)&key[0]);
        return phmap::HashState().combine(0, x, k.key[8]);
    }
```

**Comment by PietropaoloFrisoni** [comment]
I still don't fully understand how constructors work, but would it be more efficient to just create 4 keys inside the class (but outside the class method):
```
MyKey k1,k2,k3,k4;
```
and then read the hash tables as:

```
value_1= h[k1];
value_2 = h[k2];
(...)
```

ecc.. instead of creating a MyKey object every time? Or are the two things completely equivalent?

**Comment by greg7mdp** [comment]
If you use the key only once it is equivalent.
For your use case I would do this:

```

                            for (tps1 = tps1_min; tps1 <= tps1_max; tps1 += 2)
                            {
                                tps2_min = max(abs(tps1 - dspin), abs(tb3 - dspin));
                                tps2_max = min(tps1 + dspin, tb3 + dspin);

                                MyKey k3{ti_8, ti_7, ti_6, ti_5, tb1, tb2, tb3, tps1, 0};
                                MyKey k4{ti_9, ti_10, ti_11, ti_12, tb5, tb4, tb3, tps1, 0};

                                for (tps2 = tps2_min; tps2 <= tps2_max; tps2 += 2)
                                {
                                    k3[9] = k4[9] = tps2;
                                    value_3 = h[k3];
                                    value_4 = h[k4];
                                    (...);
                                } // tps2
                            }   // tps1
```

**Comment by greg7mdp** [comment]
Just updated the comment above with more key sharing

**Comment by PietropaoloFrisoni** [comment]
Thank you. 

There is just one very last thing that I don't understand. 
Each key (k1, k2, k3, k4) is used inside the method, but inserting the lines:

```
MyKey k3{ti_8, ti_7, ti_6, ti_5, tb1, tb2, tb3, 0, 0};
MyKey k4{ti_9, ti_10, ti_11, ti_12, tb5, tb4, tb3, 0, 0};
```

inside the class method should imply that these objects are created and destroyed a lot of times (the method "class_method_which_is_called_a_lot_of_times_()" is actually called at least 10^6 times in my code XD).

Would it be more efficient to just create 4 private keys:
```
MyKey k1, k2, k3, k4;
```
outside the class function and then re-initialize k1,k2,k3,k4 every time inside the function itself?

With all the help that you provided you should be a co-author in our paper hahaha

**Comment by greg7mdp** [comment]
> Would it be more efficient to just create 4 private keys outside the class function and then re-initialize k1,k2,k3,k4 every time inside the function itself?

no it wouldn't. the creation costs nothing it is just the initialization that matters

**Comment by PietropaoloFrisoni** [comment]
Edit:

It's more a curiosity that I real need (if you have other things to do there's no need for you to answer),
but with:

```
struct MyKey
{
  friend size_t hash_value(const MyKey &k)
  {
      uint64_t x = *((uint64_t *)&key[0]);
      return phmap::HashState().combine(0, x, k.key[8]);
  }

  bool operator==(const MyKey o) const { return key == o.key; }

  std::array<uint8_t, 9> key;
};
```

I get the error:
```
"a nonstatic member reference must be relative to a specific object."
```

Also, trying to do something like:

```
MyKey k3{ti_8, ti_7, ti_6, ti_5, tb1, tb2, tb3, 0, 0};
MyKey k4{ti_9, ti_10, ti_11, ti_12, tb5, tb4, tb3, 0, 0};
 
(...)
      k3[8] = k4[8] = tps1;
```

I get:

```
No operator [] corresponds to these operands (MyKey and Int)
```

**Comment by greg7mdp** [comment]
sorry, it should be:

```
struct MyKey
{
  friend size_t hash_value(const MyKey &k)
  {
      uint64_t x = *((uint64_t *)&k.key[0]);
      return phmap::HashState().combine(0, x, k.key[8]);
  }

  uint8_t& operator[](int idx) { return key[idx]; }

  bool operator==(const MyKey o) const { return key == o.key; }

  std::array<uint8_t, 9> key;
};
```

**Comment by PietropaoloFrisoni** [comment]
I already thought that the problem could be k.key[0] instead of key[0], but with the new version I keep getting this error:

```
test_sampler: ext/parallel_hashmap/phmap.h:2142: void phmap::priv::raw_hash_set<Policy, Hash, Eq, Alloc>::emplace_at(size_t, Args&& ...) [with Args = {const std::piecewise_construct_t&, std::tuple<MyKey&&>, std::tuple<>}; Policy = phmap::priv::FlatHashMapPolicy<MyKey, double>; Hash = phmap::Hash<MyKey>; Eq = phmap::EqualTo<MyKey>; Alloc = std::allocator<std::pair<const MyKey, double> >; size_t = long unsigned int]: Assertion `PolicyTraits::apply(FindElement{*this}, *iterator_at(i)) == iterator_at(i) && "constructed value does not match the lookup key"' failed.
./run_tests: riga 16:  3300 Annullato               (core dump created) bin/test_sampler

```

I left some italians words but I hope you understand what's the problem (it is completely impossible for me to figure it out xD)

**Comment by greg7mdp** [comment]
Hum, I would have to see your current code to help.
